pipeline/matching/rankers/quantity.py

```diff
@@ -6,12 +6,30 @@
 RANKER_QUANTITY = "quantityRanker"
 
 
+def get_number_matches(matches):
+    num_matches = {}
+    for match in matches:
+        for user in match.users:
+            if num_matches.get(user) is None:
+                num_matches[user] = 1
+            else:
+                num_matches[user] += 1
+    return num_matches
+
+
 class QuantityRanker(MatchRanker):
     def __init__(self):
         super().__init__(name=RANKER_QUANTITY)
 
     def rank(
         self, inp: MatchingInput, proposed: Iterator[Match]
     ) -> Iterator[Match]:
-        # TODO: Implement your ranker
-        yield from proposed
+        """The function 'rank' returns a sorted iterator based on the total number of matches for each user.
+        The less matches a user has the higher they are prioritized."""
+        num_matches = get_number_matches(proposed)
+        yield from sorted(
```

Comment:
The input to the ranking function will be a generator of matches. One gotcha with generators is that when you loop through the values, they are taken out of the generator, so the second loop will find an empty generator.

Read more about this in the [intern handbook Python tips](https://github.com/scarletstudio/handbook/wiki/Python-Tips#generators).

For now, we can handle it by doing this:

```python
# Copy generator into a list and use that list
proposed_list = list(proposed)
num_matches = get_number_matches(proposed_list)
yield from sorted(proposed_list, ...)
```

In the future, this could become memory-inefficient. Imagine we have generators that produce every possible match. If there are N users, that means N * (N - 1) matches per generator. If N is a big number, we will have N^2 matches being copied into memory, rather than being read one-by-one from a generator. At our current number of users, this is no problem, and probably won't be a big deal if communities on Butterfly are small (~100 users). But imagine if we had 1000 users and 10 generators that each produce every option, we could have 10 million matches being copied into memory!

As a challenge, you could come back to this in the future and find a way to do this without copying the generator, but that would be more for personal learning rather than an urgent need for the system.

Reply:
Implemented iterator into a list, however it is memory-inefficient.